Invert the order in which the belongsToMany joins are created

lorenzo opened the pull request:
This took me a couple months to get right... but finally here it is.

It has been a frequently reported problem the order in which the belongsToMany `matching` joins where created up until this point. For example:

``` sql
select * from articles inner join tags inner join articles_tags on (...) ...
```

The main problems have been:
1. It was not possible to reference the joint table from the query builder closure:

``` php
$query->matching('tags', function ($q) { $q->where('ArticlesTags.foo is null'); }) // SQL ERROR
```
1. `leftJoinWith` produces surprisingly odd results, although they are semantically correct.

``` sql
-- $query->leftJoinWith('tags');

LEFT JOIN tags on 1 = 1 -- This is surprising to some and don't know how to deal with it
LEFT JOIN articles_tags on (...)
WHERE articles_tags.id is null
```

The changes in this branch solve both problems by inverting the order of the joins:

``` sql
select * from articles inner join articles_tags on (...) inner join tags on (...) ...
```

These changes are potentially problematic: I'm not really sure what all the consequences for people's code might be. One example of the consequences is that I had to re-implement `notMatching`  for belongs to many associations, given that the new joins order made it impossible to keep the current logic.

In the great majority of applications, I would say that the changes are beneficial and would very rarely produce any noticeable change for anyone, but I'm targeting 3.next so that people have the chance to test before we mark it as stable.

lorenzo committed:
Inverted the order in which the joins for belongsToMany are added

lorenzo committed:
Fixing test case to reflect a more real use case

lorenzo committed:
Using another strategy for implementing notMatching in belongsToMany

After chaning the joins order, it was impossible to keep using the old way,
now it uses a NOT IN (subequery)

lorenzo committed:
Fixed the remaining failing tests

lorenzo commented on tests/TestCase/ORM/QueryTest.php:
```diff
@@ -2993,22 +2993,21 @@ public function testLeftJoinWithNested()
 
         $expected = [
-            1 => 2,
+            1 => 0,
```
I changed this test so that it does not fail in mysql 5.7 with the new `group by` mode

markstory commented on src/ORM/Association/BelongsToMany.php:
```diff
@@ -343,30 +347,61 @@ public function attachTo(Query $query, array $options = [])
+
+        $foreignKey = $this->targetForeignKey();
+        $thisJoin = $query->clause('join')[$this->name()];
+        $thisJoin['conditions']->add($assoc->_joinCondition(['foreignKey' => $foreignKey]));
```
Interesting that adding the foreign key later on helps solve the problems.

lorenzo replied:
Well, what solves problems is changing the joins order. This line is so that each of the joins share parts of the conditions. Before, the conditions where always put in the last join

markstory commented on src/ORM/Association/BelongsToMany.php:
```diff
@@ -343,30 +347,61 @@ public function attachTo(Query $query, array $options = [])
-                array_map([$exp, 'isNull'], $primaryKey);
-                return $exp;
-            });
+        if (empty($options['negateMatch'])) {
```
Is this necessary given line 334?

lorenzo replied:
Unfortunately... I can remove line 334, though

lorenzo replied:
No, it seems I still need both. The function have different call paths

markstory replied:
ok

markstory commented on src/ORM/Association/BelongsToMany.php:
```diff
@@ -343,30 +347,61 @@ public function attachTo(Query $query, array $options = [])
+
+        $subquery = $options['queryBuilder']($subquery);
+
+        $assoc = $this->junction()->association($this->target()->alias());
```
Can't you use `$junction` here?

lorenzo replied:
I can!

markstory commented on src/ORM/Association/BelongsToMany.php:
```diff
@@ -343,30 +347,61 @@ public function attachTo(Query $query, array $options = [])
+            ->andWhere(function ($exp) use ($subquery, $conds) {
+                $identifiers = [];
+                foreach (array_keys($conds) as $field) {
+                    $identifiers = new IdentifierExpression($field);
```
Does `IdentifierExpression` handle the field being a complex type, like a function expression?

lorenzo replied:
It does not, but the fields here are always strings. The reason is that each `$field` was a key in the `$conds` array.

markstory replied:
Good point. I didn't think of that.

markstory commented on tests/TestCase/ORM/QueryTest.php:
```diff
@@ -2993,22 +2993,21 @@ public function testLeftJoinWithNested()
-            ->select(['total_articles' => 'count(articles.id)'])
+            ->select([
+                'authors.id',
+                'total_articles' => 'count(tags.id)'])
```
Why is total_articles a count of tags.id?

lorenzo replied:
Misleading select alias, I guess. It is the count of articles with that particular tag

lorenzo committed:
Simplifying code a bit

lorenzo committed:
Making test more clear and re-adding missing junction conditions in subquery
for `notMatching`

lorenzo commented:
@markstory I renamed the select alias in the test for clarity and also added again the `junctionConditions()` that I missed for `notMatching()`.

I think this is good to go

GuidoHendriks commented:
After creating issue #9051 I discussed this matter with @lorenzo. The current behaviour isn’t as you would expect from `notMatching`. It should be complement to `matching`, like the documentation suggests and how it’s implemented for other relation types.

The current behaviour of `notMatching` using BTM relations definitely has it's use, this is however a complement to `contain` instead of `matching`. My suggesting would be merging this PR into 3.2 to get the behaviour of `notMatching` right. And then implement a new method for the current behaviour.

lorenzo commented:
I'm ok with merging this to master, there are more people unhappy than satisfied with the current behavior of notMatching in 3.2.

What do you think @markstory ?

markstory commented:
@lorenzo Its a moderately risky change for a bugfix release. We could end up having to triage more issues/break people's apps.

I'm really hoping to have a beta of 3.3 out next week if that helps.

GuidoHendriks commented:
I would say leaving a bad implementation in is quite risky too. Besides that the behaviour differs from BTM to other relation types, it's also different based on the fact if there are one or multiple matches for the conditions.

ADmad commented:
@GuidoHendriks While it may be different people might be relying on that different behavior. In a bugfix release ensuring there are no regressions is more important then changing/fixing something.

GuidoHendriks commented:
Since it's behaviour is not only unexpected but also inconsistent (1 match returns filtered entities once, multiple matches return all entities multiple times), I would say it could cause more harm than good.  But that's my opinion, it's completely up to you guys.

This behaviour is probably worth mentioning in the documentation, since the behaviour is different that what the documentation claims.

I will just keep using my workaround, which just users the proper joins like implemented in this PR.

ionas commented:
1. In case it does not get merged 👍 for making a fat notice in the docs AND a guide how to circumvent the issue.
2. In case it does get merged there should be a **big red warning** in the changelog.
3. I wonder if "4.0-soon" could also fix it - like, let HTTP Middleware proofs itself and then go for 4.x? A major release every 1-2 years should not hurt? After all the point of semver is to stick with it (usually)?

> My suggesting would be merging this PR into 3.2 to get the behaviour of notMatching right. And then implement a new method for the current behaviour.
Could `notMatching()` stay and once you add a flag while calling it you get "the new behavior"?

Thanks for your work on this issue @lorenzo and @GuidoHendriks
